pkg/controllers/cinder/cinder.go

```diff
@@ -0,0 +1,151 @@
+// from Cinder.
+
+type CinderController struct {
+	operatorClient     v1helpers.OperatorClient
+	kubeClient         kubernetes.Interface
+	openStackClient    *openStackClient
+	storageClassLister storagelisters.StorageClassLister
+	// Controllers to start when Cinder is detected
+	csiControllers     []Runnable
+	controllersRunning bool
+	eventRecorder      events.Recorder
+}
+
+type Runnable interface {
+	Run(ctx context.Context, workers int)
+}
+
+const (
+	// Minimal interval between controller resyncs. The controller will detect
+	// new share types in Cinder and create StorageClasses for them at least
+	// once per this interval.
+	resyncInterval = 20 * time.Minute
+
+	operatorConditionPrefix = "CinderController"
+)
+
+func NewCinderController(
+	operatorClient v1helpers.OperatorClient,
+	kubeClient kubernetes.Interface,
+	informers v1helpers.KubeInformersForNamespaces,
+	openStackClient *openStackClient,
+	csiControllers []Runnable,
+	eventRecorder events.Recorder) factory.Controller {
+
+	scInformer := informers.InformersFor("").Storage().V1().StorageClasses()
+	c := &CinderController{
+		operatorClient:     operatorClient,
+		kubeClient:         kubeClient,
+		storageClassLister: scInformer.Lister(),
+		openStackClient:    openStackClient,
+		csiControllers:     csiControllers,
+		eventRecorder:      eventRecorder.WithComponentSuffix("CinderController"),
+	}
+	return factory.New().WithSync(c.sync).WithSyncDegradedOnError(operatorClient).ResyncEvery(resyncInterval).WithInformers(
+		operatorClient.Informer(),
+		scInformer.Informer(),
+	).ToController("CinderController", eventRecorder)
+}
+
+func (c *CinderController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	klog.V(4).Infof("Cinder sync started")
+	defer klog.V(4).Infof("Cinder sync finished")
+
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorv1.Managed {
+		return nil
+	}
+
+	volumeTypes, err := c.openStackClient.GetVolumeTypes()
+	if err != nil {
+		return err
+	}
+
+	if len(volumeTypes) == 0 {
+		err = fmt.Errorf("no volume types found for Cinder service")
+		klog.V(2).Error(err, "No volume types found for Cinder service")
+		return err
+	}
+
+	if !c.controllersRunning {
+		klog.V(4).Infof("Starting CSI driver controllers")
+		for _, ctrl := range c.csiControllers {
+			go ctrl.Run(ctx, 1)
+		}
+		c.controllersRunning = true
+	}
+	err = c.syncStorageClasses(ctx, volumeTypes)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *CinderController) syncStorageClasses(ctx context.Context, volumeTypes []volumetypes.VolumeType) error {
+	var errs []error
+	for _, volumeType := range volumeTypes {
+		klog.V(4).Infof("Syncing storage class for volumeType type %s", volumeType.Name)
+		sc := c.generateStorageClass(volumeType)
+		_, _, err := resourceapply.ApplyStorageClass(c.kubeClient.StorageV1(), c.eventRecorder, sc)
```

Comment:
Right now, `ApplyStorageClass` will try to update also immutable field, which is going to fail. This error will be returned up the stack and can make the whole operator `Degraded`. IMO, logging the errors could be enough. Or you can re-create the storage class (delete + create).